File: ui/input_state.py
"""InputState: an immutable snapshot of input for one frame.

The UI module accumulates raw GLFW events as they arrive (asynchronously, via
callbacks) and freezes them into one of these at the top of each frame. Everything
downstream reads the snapshot rather than polling GLFW, so every consumer in a
frame sees exactly the same input -- no mid-frame tearing where the physics step
and the UI disagree about where the mouse is.

CAPTURE IS ALREADY RESOLVED IN HERE. The `*_captured` fields record what imgui
claimed, and the plain fields (`left_pressed`, `scroll`, `keys_pressed`, ...) are
already filtered: they describe input meant for the CANVAS. A consumer that wants
canvas input just reads the plain field and does not think about imgui at all.
The unfiltered variants exist for the rare case that needs them (e.g. dismissing
a transient overlay on any click, wherever it landed).
"""

from __future__ import annotations

from dataclasses import dataclass, field


@dataclass(frozen=True)
class InputState:
    """One frame's input. Frozen: consumers read, never mutate."""

    # --- pointer ---------------------------------------------------------
    # Screen pixels, GLFW convention (origin top-left, y down).
    mouse_pos: tuple[float, float] = (0.0, 0.0)
    mouse_prev: tuple[float, float] = (0.0, 0.0)
    #: Movement since last frame. Zero on the frame the mouse enters the window,
    #: so a re-entering cursor cannot produce a huge phantom drag.
    mouse_delta: tuple[float, float] = (0.0, 0.0)

    #: Held state, filtered by imgui capture.
    left_held: bool = False
    right_held: bool = False
    middle_held: bool = False

    #: One-shot: went down this frame (canvas only).
    left_pressed: bool = False
    right_pressed: bool = False
    middle_pressed: bool = False

    #: One-shot: came up this frame. NOT capture-filtered -- a release must
    #: always be delivered to whoever saw the press, or a drag that ends over
    #: an imgui window would never terminate and the canvas would stay "grabbed".
    left_released: bool = False
    right_released: bool = False
    middle_released: bool = False

    #: True while a canvas drag is in progress: the press landed on the canvas
    #: and the button has not come up yet. Stays true even if the cursor moves
    #: over an imgui window mid-drag, which is what makes dragging feel right.
    left_dragging: bool = False
    right_dragging: bool = False

    #: Accumulated wheel movement this frame (canvas only).
    scroll: float = 0.0

    # --- keyboard --------------------------------------------------------
    #: GLFW key codes currently held (canvas only).
    keys_held: frozenset[int] = frozenset()
    #: Went down this frame, including auto-repeat (canvas only).
    keys_pressed: frozenset[int] = frozenset()
    #: Came up this frame. Not capture-filtered, for the same reason as
    #: mouse releases.
    keys_released: frozenset[int] = frozenset()
    #: GLFW modifier bitmask from the most recent key event.
    mods: int = 0

    # --- imgui capture ---------------------------------------------------
    #: imgui wants the pointer (cursor is over a window/control).
    mouse_captured: bool = False
    #: imgui wants the keyboard (a text field has focus).
    keyboard_captured: bool = False

    #: Unfiltered one-shots: true even when imgui captured the click.
    any_left_pressed: bool = False
    any_right_pressed: bool = False

    # --- frame -----------------------------------------------------------
    #: Seconds since the previous frame.
    dt: float = 0.0

    def key_held(self, key: int) -> bool:
        return key in self.keys_held

    def key_pressed(self, key: int) -> bool:
        return key in self.keys_pressed

File: src/ui/inputState.ts
/**
 * InputState: one frame's input, frozen.
 * The type half of `ui/input_state.py` (86 lines, 24 fields).
 *
 * ## Who fills this in
 *
 * Written in Step 7 as the TYPE the frame loop is coded against --
 * `applyCanvasInput` and `applyCameraKeys` are ports of real desktop methods
 * and would otherwise have had nothing to read. **Step 8 built the producer**,
 * and split it in two so that the half with the decisions in it is testable
 * without a DOM:
 *
 *   - `inputTracker.ts` -- accumulates events and freezes this snapshot. Pure;
 *     never touches `window`. The port of `ui.py:82-255`.
 *   - `inputBinding.ts` -- the DOM listeners, which only translate events into
 *     tracker calls. The port of `ui.py:141-190`.
 *   - `hotkeys.ts` -- the focus-aware table, read from `keysPressed` below.
 *
 * ## Rebuilt once per frame, never mutated
 *
 * The desktop's is a frozen dataclass rebuilt each frame, so **every consumer
 * within a frame sees identical input**. `readonly` is the port of that, and it
 * matters more than it looks: `applyCanvasInput` and `status()` both read
 * `mousePos`, and a value that could change between them would put the readout
 * and the pick at different pixels.
 *
 * ## The asymmetries Step 8 must keep
 *
 * Stated here rather than in Step 8's handler because they are properties of
 * what these FIELDS MEAN, and a handler written without them produces fields
 * that are subtly the wrong thing:
 *
 *   - **Releases are never capture-filtered.** A button that went down on the
 *     canvas must be able to come up over a panel, or the drag never ends.
 *   - **A drag belongs to whoever received the press.** `leftDragging` stays
 *     true while the cursor wanders over the UI. That is why the drawing and
 *     shove tools read `*Dragging` rather than `*Held`.
 *   - **Capture is resolved ONCE, at the event handler.** By the time input
 *     reaches here, plain fields already mean "meant for the canvas". **No
 *     consumer downstream checks a capture flag** -- if you find yourself
 *     wanting to, the filtering belongs upstream.
 */

/** One frame's input, already filtered for UI capture. */
export interface InputState {
  /** Cursor position in framebuffer pixels, top-left origin. */
  readonly mousePos: readonly [number, number];
  /** Seconds since the previous frame. Zero on the first frame. */
  readonly dt: number;

  /** Went down THIS frame, on the canvas. One-shot. */
  readonly leftPressed: boolean;
  readonly rightPressed: boolean;
  /**
   * A drag owned by the canvas is in progress.
   *
   * NOT the same as "the button is down": a press that landed on a panel never
   * starts one, and a drag that began on the canvas survives the cursor
   * crossing a panel.
   */
  readonly leftDragging: boolean;
  readonly rightDragging: boolean;

  /**
   * Scroll notches this frame, positive up. Zero when the wheel did not move.
   *
   * A NUMBER, not a boolean plus a direction, because `zoomAtPixel` takes
   * notches and a fast flick is worth more than one.
   */
  readonly scroll: number;

  /**
   * Physical key codes currently held (`KeyW`, `KeyA`, ...).
   *
   * **`KeyboardEvent.code`, not `.key`** -- the port reads WASD as physical
   * positions, so the same keys work on AZERTY and Dvorak. `.key` would give
   * `z` where `w` sits on a French layout, and the pan would go sideways.
   *
   * HELD, not pressed: continuous motion for as long as the key is down. See
   * `applyCameraKeys` for why these deliberately bypass the hotkey table.
   */
  readonly keysHeld: ReadonlySet<string>;

  /**
   * Physical key codes that went down THIS frame. One-shot, canvas only.
   *
   * **Auto-repeat IS included**, matching the desktop: `ui.py:154-156` treats
   * GLFW's PRESS and REPEAT identically, so a held key re-enters this set every
   * repeat tick. That is harmless for the one-shots the hotkey table dispatches
   * (holding `R` re-resets, which is what a held reset key should do) and it is
   * exactly why continuous motion reads `keysHeld` instead -- repeat RATE is an
   * OS setting, so panning through this set would move at a speed the app does
   * not control.
   */
  readonly keysPressed: ReadonlySet<string>;

  /**
   * Shift, from the most recent key event OR canvas press. Trimmed port of `mods`.
   *
   * Only Shift, because the hotkey table is deliberately Ctrl-free (see
   * `hotkeys.ts`) and Shift is the only modifier any binding discriminates.
   * The desktop carries a full GLFW bitmask; porting one would mean four fields
   * nothing reads.
   *
   * **POINTER PRESSES WRITE IT TOO, and that is not redundant with the key
   * events.** Shift+Right-click is redo (`applyCanvasInput`), and deciding that
   * from the last KEY event would misread a modifier pressed after the pointer
   * went down, or one whose keydown was swallowed by a focused text field.
   * `onPointerDown` records what the browser reports at the click itself.
   */
  readonly shift: boolean;
}

/**
 * No input at all. The first frame's value, and what a headless driver passes.
 *
 * `dt: 0` is load-bearing rather than a placeholder: `applyCameraKeys` returns
 * early on a non-positive `dt`, so a frame with no measured delta cannot move
 * the camera by an unscaled step. The desktop guards the same way
 * (`orchestrator.py:447-448`).
 */
export const EMPTY_INPUT: InputState = Object.freeze({
  mousePos: Object.freeze([0, 0]) as readonly [number, number],
  dt: 0,
  leftPressed: false,
  rightPressed: false,
  leftDragging: false,
  rightDragging: false,
  scroll: 0,
  keysHeld: Object.freeze(new Set<string>()),
  keysPressed: Object.freeze(new Set<string>()),
  shift: false,
});
